src/express-secure-handlebars.js

```diff
@@ -20,20 +23,123 @@ function ExpressSecureHandlebars(config) {
 
     /* calling super constructor */
     this.constructor.super_.call(this, config);
+
+    // dedicate a special express-handlebars for fetching raw partials
+    this._exphbsForRawPartials = new ExpressHandlebarsForRawPartials(config);
+    this._partialsCache = {
+        preprocessed: {}
+    };
+
+    // compilerOptions is being used for passing params to secure-handlebars
+    this.compilerOptions || (this.compilerOptions = {});
+    this.compilerOptions.shbsPartialsCache = this._partialsCache;
 }
 
 /* inheriting the express-handlebars */
-util.inherits(ExpressSecureHandlebars, expressHandlebars);
+util.inherits(ExpressSecureHandlebars, ExpressHandlebars);
+
+// always returns an empty cache, since the partial is loaded and analyzed on-demand
+// TODO: preprocess all templates, and export all partials for compatibility
+ExpressSecureHandlebars.prototype.getPartials = function (options) {
+    return {};
+};
+
+// retrieve file content directly from the preprocessed partial cache if it exists
+ExpressSecureHandlebars.prototype._getFile = function (filePath, options) {
+
+    var file = this._partialsCache.preprocessed[path.relative('.', filePath)];
+    return file ? Promise.resolve(file) : ExpressHandlebars.prototype._getFile.call(this, filePath, options);
+
+};
+
+
+// _getTemplates literally has the core of getTemplates, except that it takes an array of filePaths as input
+/*
+// getTemplates can be rewritten to use _getTemplates in the following way
+ExpressSecureHandlebars.prototype.getTemplates = function (dirPath, options) {
+    options || (options = {});
+    
+    return this._getDir(dirPath, {cache: options.cache}).map(function (filePath) {
+        return path.join(dirPath, filePath);
+    }).then(function(filePaths) {
+        return this._getTemplates(filePaths, options);
+    }.bind(this));
+};
+*/
+ExpressSecureHandlebars.prototype._getTemplates = function (filePaths, options) {
+    options || (options = {});
+
+    var templates = filePaths.map(function (filePath) {
+        return this.getTemplate(filePath, options);
+    }, this);
+
+    return Promise.all(templates).then(function (templates) {
+        return filePaths.reduce(function (hash, filePath, i) {
+            hash[filePath] = templates[i];
+            return hash;
+        }, {});
+    });
+};
+
 
-/* override ExpressHandlebars.render() to expose filePath as compilerOptions */
+// override render() for on-demand partial preprocessing and (pre-)compilation
 ExpressSecureHandlebars.prototype.render = function (filePath, context, options) {
```

Comment:
same as SH, we should enable this new feature through configuration, and set it as false before it is stable.